Features/task 2

LindaLangerova opened the pull request.

LindaLangerova committed:
KA-242 Extend app structure by data and contract layer

KA-242 Move models to contract layer

LindaLangerova committed:
KA-242 Add repository interface and class

LindaLangerova committed:
KA-242 Move ItemList to repository

LindaLangerova committed:
KA-242 Implement all methods

LindaLangerova committed:
KA-242 Apply code cleanup

KA-242 Remove unsightly space

KA-242 Apply code cleanup

KA-242 Apply code cleanup

KA-242 Remove empty folders

KA-242 Add formating file and apply code cleanup

LindaLangerova committed:
KA-242 Add Unity dependencies

KA-242 Change Unity version and update dependencies

LindaLangerova committed:
KA-242 Add lifetime manager to type registration

LindaLangerova committed:
KA-242 Extend controller methods

LindaLangerova committed:
KA-242 Mock repository in tests

KA-242 Make mocking less universal

LindaLangerova committed:
KA-242 Reorganize solution structure

LindaLangerova committed:
KA-242 Rename WebApiConfig.cs to be named as it should be

LindaLangerova committed:
KA-242 Symplify ItemRepository.cs and move its interface to Contract

LindaLangerova committed:
KA-242 Remove app.config from places where it's not needed

LindaLangerova committed:
KA-242 Update UnityConfig to just register component and move dependency resolver to Global.asax

LindaLangerova committed:
KA-242 Repair tests

LindaLangerova committed:
KA-242 Implement UnityBootstrapper

LindaLangerova committed:
KA-242 Update routing

KA-242 Use ItemUrlManager to get item route

LindaLangerova committed:
KA-242 Register UrlManager interface by unity and update Request gathering

PetrSvirak commented on TodoApp/src/TodoApp.Api/ApiUnityBootstrapper.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public class ApiUnityBootstrapper: IUnityBootstrapper
+    {
+        public static void RegisterComponents(IUnityContainer container)
```
This static method actually beats the purpose of the `IUnityInterface` quite a bit. Get rid of it in favor of the instance method. Then you simply instantiate the classes and call their `RegisterComponents` methods.

PetrSvirak commented on TodoApp/src/TodoApp.Api/ApiUnityBootstrapper.cs:
```diff
@@ -0,0 +1,25 @@
+                .RegisterType<HttpRequestMessage, HttpRequestMessage>(new InjectionConstructor());
+        }
+
+        void IUnityBootstrapper.RegisterComponents(IUnityContainer container)
```
Consider renaming the method to something more related to either types or contracts - this is the only place you speak about `components` at any rate.

PetrSvirak commented on TodoApp/src/TodoApp.Api/App_Start/UnityConfig.cs:
```diff
@@ -0,0 +1,20 @@
+            IUnityContainer container = new UnityContainer();
+
+            DataUnityBootstrapper.RegisterComponents(container);
+            ApiUnityBootstrapper.RegisterComponents(container);
```
Once you rewrite this to the instance method calls, you should see some repeating pattern here. Try to decrees the "same type of code" repetition by extracting the pattern into a private method - leverage your knowledge of fluent API pattern and staticity of the class (sorry for this quizzy comment, but you will learn more from inventing the solution that absorbing someone else's one I believe).

PetrSvirak commented on TodoApp/src/TodoApp.Api/App_Start/UnityConfig.cs:
```diff
@@ -0,0 +1,20 @@
+            DataUnityBootstrapper.RegisterComponents(container);
+            ApiUnityBootstrapper.RegisterComponents(container);
+
+            GlobalConfiguration.Configuration.DependencyResolver = new UnityDependencyResolver(container);
```
Always use `GlobalConfiguration.Configure`, ideally in the `global.asax.cs` - there is a pretty nice pattern you already use with `RouteConfig`, please stick to it.

PetrSvirak commented on TodoApp/src/TodoApp.Api/ItemUrlManager.cs:
```diff
@@ -0,0 +1,24 @@
+
+        public ItemUrlManager(UrlHelper urlHelper)
+        {
+            urlHelper.Request = (HttpRequestMessage)HttpContext.Current.Items["MS_HttpRequestMessage"];
```
Why? 😞
First, it will work even without this line.
Second, it goes against `DependencyInjection` pattern and renders your code untestable.

PetrSvirak commented on TodoApp/src/TodoApp.Api/ItemUrlManager.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace TodoApp.Api
+{
+    public class ItemUrlManager:IItemUrlManager
```
You have no tests for this class. You really should stick with the TDD:
* create an interface (`IItemUrlManager`)
* create its implementation that throws NotImplementedException (`ItemUrlManager`)
* write a test that calls `GetItemUrl` with a specific `Guid` and expects it to return  an URL
* change the code in the most trivial way to work (returning a string with inlined `id` is OK)
* when the test passes, make a new one (or rewrite the test - it is not by the book, but you might find the early tests cluttering later in the process) that sets a mock for `UrlHelper` and expects the result to match the mock - it is OK you do not have a place to pass the mock in, you know you intend the class to be aligned with predefined routes.
* change the code so it accepts `UrlHelper` in its constructor - you cannot pass it to the method since you mustn't change the contract - and call it in order to make the tests pass again
* eventually refactor the code making sure the tests still pass
* eventually refactor the tests while not changing code, making sure there are no redundant tests

PetrSvirak commented on TodoApp/src/TodoApp.Api/ItemUrlManager.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        private readonly UrlHelper _urlHelper;
+
+        public ItemUrlManager(UrlHelper urlHelper)
```
Let's thing about something else than `-Manager` - suffixes like `Manager`, `Helper`, or `Utilities` are usually a sign of the [naming code smell](https://blog.codinghorror.com/i-shall-call-it-somethingmanager/).

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Controllers/ItemListController.test.cs:
```diff
@@ -19,13 +21,35 @@ public class ItemListControllerTest : TestBase
             };
         }
 
+        internal IItemRepository MockItemRepository()
```
It can be private method, can't it?

PetrSvirak replied:
At any rate, move it to the bottom of the class, please.

PetrSvirak commented on TodoApp/src/TodoApp.Api/ApiUnityBootstrapper.cs:
```diff
@@ -0,0 +1,25 @@
+        {
+            container
+                .RegisterType<IItemUrlManager, ItemUrlManager>(new HierarchicalLifetimeManager())
+                .RegisterType<UrlHelper, UrlHelper>(new HierarchicalLifetimeManager())
```
You actually don't need to register this one, but it is OK to leave it here - it is JFYI. Also, find out how `Unity` chooses what constructor will it use.

PetrSvirak commented on TodoApp/src/TodoApp.Contract/IItemUrlManager.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace TodoApp.Contract
```
While having `IUnityBootstrapper` in the root of `Contracts` is justifiable (since all entry points of other assemblies implement it), it is definitely not the case of `IItemUrlManager` - make a folder for it.

PetrSvirak replied:
This also applies to its implementations...

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Utilities/TestItemUrlManager.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace TodoApp.Api.Tests.Utilities
+{
+    class TestItemUrlManager: IItemUrlManager
```
Simply use `NSubstitute` to do this work for you 😉

PetrSvirak commented on TodoApp/src/TodoApp.Api/ItemUrlManager.cs:
```diff
@@ -0,0 +1,24 @@
+
+        public string GetItemUrl(Guid id)
+        {
+            return _urlHelper.Route("DefaultApi", new {id});
```
What does this hardcoded string doing here? 😉 
This definitely is a constant. I'm not sure why you reverted to `convention based routing`, however, it will definitely not save you from solving this "puzzle" - no matter what routing you use, you will have to admit to yourself that this route name is `ItemUrlManager`'s dependency… 😛

PetrSvirak commented on TodoApp/src/TodoApp.Contract/Models/ItemModel.cs:
```diff
@@ -1,6 +1,6 @@
-namespace TodoApp.Api.Models
+namespace TodoApp.Contract.Models
 {
     public class ItemModel
```
Please, get rid of the `Model` suffix here (unless you are too much personally attached to the name 😃)

PetrSvirak reviewed (changes requested):
The code seems to be quite OK. Let's not push more complexity into it just now - we can save it for task 3 since it might be too early to discuss Unity isolation or API's service layer. Other than complexity, there are some essential parts of your code that need further attention from you described below... And of course a detail or two.

In general, I would like you to use more expression bodied functions - it shortens the code and makes it more readable when applied correctly.

I have also noticed some disproportionalities in the way you write similar code at different places across your solution (most mischievous was `<Class>:<Interface>` - do you see spaces around the colon? So don't I 😆).

Both problems above have the same solution: Tinker with R#'s configuration regards code style (`Formatting style`, `File layout`, `Code style`) to write your code with quite a style and in a daring format by pressing just 3 buttons (`CTRL` + `E`, `C`). Of course, don't forget to commit these project-wide R# settings to your repository for others (well, me 👼) to enjoy 😉 I believe that Rado has some fresh experience he can share with You.

PetrSvirak commented:
For an unknown reason, some comments were left out. I will thus post them separately below with links to the related lines of code (sorry for the inconvenience):

PetrSvirak commented:
[ItemUrlManager, L9](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-ac8c2c8b1dd3289e49a824ad7f5c251eR9):
You have no tests for this class. You really should stick with the TDD:
* create an interface (`IItemUrlManager`)
* create its implementation that throws NotImplementedException (`ItemUrlManager`)
* write a test that calls `GetItemUrl` with a specific `Guid` and expects it to return  an URL
* change the code in the most trivial way to work (returning a string with inlined `id` is OK)
* when the test passes, make a new one (or rewrite the test - it is not by the book, but you might find the early tests cluttering later in the process) that sets a mock for `UrlHelper` and expects the result to match the mock - it is OK you do not have a place to pass the mock in, you know you intend the class to be aligned with predefined routes.
* change the code so it accepts `UrlHelper` in its constructor - you cannot pass it to the method since you mustn't change the contract - and call it in order to make the tests pass again
* eventually refactor the code making sure the tests still pass
* eventually refactor the tests while not changing code, making sure there are no redundant tests

PetrSvirak commented:
[ItemUrlManager, L13](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-ac8c2c8b1dd3289e49a824ad7f5c251eR13):
Let's thing about something else than `-Manager` - suffixes like `Manager`, `Helper`, or `Utilities` are usually a sign of the [naming code smell](https://blog.codinghorror.com/i-shall-call-it-somethingmanager/).

PetrSvirak commented:
[ItemListControllerTest, L32](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-aab21924f62fbbc571759f5b26ebe5b4R32):
It can be private method, can't it?

PetrSvirak commented:
[ItemListControllerTest, L32](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-aab21924f62fbbc571759f5b26ebe5b4R32):
At any rate, move it to the bottom of the class, please.

LindaLangerova committed:
KA-242 Rename ItemModel to be just Item

LindaLangerova committed:
KA-242 Make RegisterComponents instance method instead of static one

LindaLangerova committed:
KA-242 Rename method

LindaLangerova committed:
KA-242 Extract RegisterTypes code repetition into the separated mehod

LindaLangerova committed:
KA-242 move unity configuration to global.asax

LindaLangerova committed:
KA-242 Remove Request gathering

LindaLangerova committed:
KA-242 Rename ItemUrlManager to raise its levelheadedness

LindaLangerova committed:
KA-242 Create ItemUrlObtainerTest

LindaLangerova committed:
KA-242 Mock instead of using TestItemUrlObtainer

LindaLangerova committed:
KA-242 Move to Utilities directory

LindaLangerova committed:
KA-242 Differ colliding folder name

LindaLangerova committed:
KA-242 Update .gitignore

LindaLangerova committed:
KA-242 Change name

LindaLangerova committed:
KA-242 Beautify UnityConfig code

LindaLangerova committed:
KA-242 Update mocking

LindaLangerova committed:
KA-242 Apply Renaming

KA-242 Rename components and complete tests

KA-242 Rename UrlGenerator

LindaLangerova committed:
KA-242 Make UrlGenerator more independent

PetrSvirak commented on TodoApp/src/TodoApp.Api/Controllers/ItemListController.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace TodoApp.Api.Controllers
+{
+    [ApiVersion("1.0")] public class ItemListController : ApiController
```
Place each attribute on its own line, please

PetrSvirak replied:
The controller name might be quite descriptive on its own, however, I believe it is not much [RESTful](https://restfulapi.net/resource-naming/)…

PetrSvirak commented on TodoApp/src/TodoApp.Api/App_Start/RouteConfig.cs:
```diff
@@ -0,0 +1,27 @@
+            var versionConstraintResolver = new DefaultInlineConstraintResolver
+            {
+                ConstraintMap =
+                    {["apiVersion"] = typeof(ApiVersionRouteConstraint)}
```
Please, set you R# up not to wrap so early. You will [save some extra lines](https://imgflip.com/i/29n0nk).

PetrSvirak replied:
Here are two options for the code alignment I see:
1. move the `{` to the same line as `=` – expecting no new constraints (preferable in your case)
1. move the `{` and `}` to stand-alone lines, place a comma behind `[…] = typeof(…)` – expecting lots of constraints

PetrSvirak commented on TodoApp/src/TodoApp.Api/Services/UrlGenerator.cs:
```diff
@@ -0,0 +1,18 @@
+
+        public string GetItemUrl(Guid id)
+            => _urlHelper.Route("DefaultApi", new
+                                    {id});
```
Seriously? 😛 \[[lines](https://imgflip.com/i/29n1c6)]

PetrSvirak commented on TodoApp/src/TodoApp.Contract/Repositories/IItemRepository.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace TodoApp.Contract.Repositories
+{
+    public interface IItemRepository
```
Make it async, please. 

Repositories simply don't sync. Since we discussed this topic more generally and in greater details and I promised you some links, spend some time with them, please. I found two SOs and MS doc that might intrigue you: [SO](https://stackoverflow.com/a/29809054/1138663) [SO](https://stackoverflow.com/a/29809119/1138663) Microsoft](https://docs.microsoft.com/en-us/dotnet/csharp/async).

PS: Not seeing 3 mildly-funny puns urges you to read the comment again 😛

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Controllers/ItemListController.test.cs:
```diff
@@ -0,0 +1,172 @@
+
+            itemRepository.GetAll().Returns(ReferencedItems.Clone());
+            itemRepository.Get(fakeId).Returns(fakeItem);
+            itemRepository.Add(Arg.Any<Item>()).Returns(fakeItem);
```
Using `Arg.Any<T>()` might lead to false-positive tests. For example, passing the `null` into the `repository.Add` method [here](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-61318171ce5351894b247b4cc7350f16R34) will not fail your `[PostNewItem_UniqueItem_ItemAdded](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-aab21924f62fbbc571759f5b26ebe5b4R130)` test.

PetrSvirak replied:
It is not unusual to configure substitutes in the body of the test that requires them to work in a given way. In your case, each method you fake is used in only test. If you make `itemRepository` a class field, you can tailor its configuration for each individual test. The only thing that would remain of the `MockItemRepository` would be the [`Substitute` line](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-aab21924f62fbbc571759f5b26ebe5b4R65)

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/ItemUrlObtainer.test.cs:
```diff
@@ -0,0 +1,40 @@
+
+            var receivedUrl = _itemUrlGenerator.GetItemUrl(id);
+
+            Assert.That(receivedUrl, Is.EqualTo(requestedUrl));
```
This test passes if I replace `GetItemUrl`'s implementation with `=> urlHelper.Route("DefaultApi", new {})`.

PetrSvirak replied:
Mind you, once you fix the `Any` in `SetUp`, there is one more flaw: test written in the way it is now will pass if I replace your `GetItemUrl`'s implementation with `urlHelper.Route("DefaultApi", new {}) + id` – do check :)

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/ItemUrlObtainer.test.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace TodoApp.Api.Tests.Services
+{
+    internal class ItemUrlGeneratorTest : TestBase
```
What about the name of the class? 😛

LindaLangerova replied:
:D Oh my... it's 3 in 1! -.-

PetrSvirak commented on TodoApp/src/TodoApp.Api/App_Start/RouteConfig.cs:
```diff
@@ -0,0 +1,27 @@
+            config.MapHttpAttributeRoutes(versionConstraintResolver);
+
+            config.Routes.MapHttpRoute(
+                "DefaultApi",
```
You need this string to match to [that string](https://github.com/LindaLangerova/kentico-onboarding-cs/pull/3/files#diff-ac8c2c8b1dd3289e49a824ad7f5c251eR19) in `UrlGenerator` – always… What a constant phenomenon 😆

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/TodoApp.Api.Tests.csproj:
```diff
@@ -0,0 +1,178 @@
+    <Reference Include="Unity.ServiceLocation, Version=2.0.3.0, Culture=neutral, PublicKeyToken=489b6accfaf20ef0, processorArchitecture=MSIL">
+      <HintPath>..\..\packages\Unity.5.2.0\lib\net45\Unity.ServiceLocation.dll</HintPath>
+    </Reference>
+  </ItemGroup>
```
Wow. So many references. You mustn't catch them all 😉 \[[R# to the rescue](https://www.jetbrains.com/help/resharper/Navigation_and_Search__Finding_Usages__Optimizing_References.html)]

PetrSvirak reviewed (changes requested):
It nice and almost clean - cool 😺 There are but few comments below, however, almost all of them will require you to go through your whole solution making proposed changes. 

I also copied the links related to our async/await discussion here, so, please spend some work-time dwelling upon the topic.

Happy hacking😃

LindaLangerova committed:
KA-242 Cleanup code

LindaLangerova committed:
KA-242 Rename few classes

ItemListController → ItemsController
ItemListControllerTest → ItemsControllerTest
ItemUrlGeneratorTest → UrlGeneratorTest

LindaLangerova committed:
KA-242 Make repository async and update async flow in controller and tests

LindaLangerova committed:
KA-242 Make default url a constant and update test conditions

LindaLangerova committed:
KA-242 Optimize references and cleanup code

LindaLangerova committed:
KA-242 Test more specific url

PetrSvirak commented on TodoApp/src/TodoApp.Api/App_Start/RouteConfig.cs:
```diff
@@ -11,17 +11,13 @@ public static void Register(HttpConfiguration config)
-                new
-                    {id = RouteParameter.Optional});
+            config.Routes.MapHttpRoute("DefaultApi", "api/v{version:apiVersion}/{controller}/{id}",
+                                       new {id = RouteParameter.Optional});
```
If you decide to break lines on parameter bases, you should break after every single parameter you pass. Multiple arguments won't blend together this way (just like the `"DefaultApi"` and `path` in this case).

PetrSvirak replied:
I see no R# template update around – at least definition of "too long line" can be altered so reformatting does not break (line) your code…

PetrSvirak commented on TodoApp/src/TodoApp.Data/Repositories/ItemRepository.cs:
```diff
@@ -15,17 +16,17 @@ public class ItemRepository : IItemRepository
 
-        public IEnumerable<Item> GetAll()
-            => ItemList;
+        public async Task<Item[]> GetAll()
```
There should be the `Async` suffix in all theses methods.

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/UrlGenerator.test.cs:
```diff
@@ -8,7 +8,7 @@
 namespace TodoApp.Api.Tests.Services
 {
-    internal class ItemUrlGeneratorTest : TestBase
+    internal class UrlGeneratorTest : TestBase
```
You should use the plural with your test classes (e.g. `UrlGeneratorTests` – see the little `s` at the end)

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/TodoApp.Api.Tests.csproj:
```diff
@@ -127,9 +127,9 @@
+    <Compile Include="Services\UrlGenerator.test.cs" />
     <Compile Include="Utilities\ActionsResolution\ActionsResolutionExtensions.cs" />
-    <Compile Include="Controllers\ItemListController.test.cs" />
+    <Compile Include="Controllers\ItemsController.test.cs" />
```
Always name your files the same way you name classes, please (here, it would be `ItemsControllerTests.cs`)

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Controllers/ItemsController.test.cs:
```diff
@@ -91,16 +91,15 @@ public void SetUp()
 
             var fakeId = Guid.Parse("c5cc89a0-ab8d-4328-9000-3da679ec02d8");
-            var fakeItem = new Item { Id = fakeId, Text = "Make a coffee 2" };
+            var fakeItem = new Item {Id = fakeId, Text = "Make a coffee 2"};
```
No R# template update? 😛

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/UrlGenerator.test.cs:
```diff
@@ -10,29 +10,26 @@ namespace TodoApp.Api.Tests.Services
+        private UrlGenerator _itemUrlGenerator;
 
-            _itemUrlGenerator = new UrlGenerator(urlHelper);
+        private static string AnonymousArg(object first, string argName)
```
Private methods should be below public ones.

PetrSvirak replied:
This is not a name for a `method` but for a `class`.

To get more on naming read through first few chapters of [Clean Code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) or visit some discussions on the topic - such as [this one](https://softwareengineering.stackexchange.com/questions/119345/meaningful-concise-method-naming-guidelines)

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/UrlGenerator.test.cs:
```diff
@@ -10,29 +10,26 @@ namespace TodoApp.Api.Tests.Services
+        private static string AnonymousArg(object first, string argName)
+        {
+            new RouteValueDictionary(first).TryGetValue(argName, out var firstId);
+            return firstId?.ToString();
```
You really convert a `GUID` to `string` for no reason. 
* Change method signature to accept `object` and return boolean
* Compare the first argument from the `RouteValueDictionary` with a hardcoded `Guid` 
* Use the method delegate (name) in the `Arg.Is` call

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Controllers/ItemsController.test.cs:
```diff
@@ -118,8 +115,13 @@ private static IItemRepository MockItemRepository()
-            var updateItem = new Item {Id = id, Text = "Make a coffee"};
+            _repository
+                .Update(id,
+                        Arg.Is<Item>(item => item.Id.ToString() == "c5cc89a0-ab8d-4328-9000-3da679ec02d3" &&
```
If you compare two `item` you should always use comparer so you have the comparison logic at one place.

PetrSvirak replied:
Moreover, this call is really long - always introduce a method in case your lambdas grow long or complex.

PetrSvirak reviewed (changes requested):
I have initially wanted you to fix the details into the next task, however, there are some essentials that are part of this task, so let's give this task one more round 😼 

Before you get to individual issues, please, take some time to reconcile with your commits (so your future commits are better organized):
* 45c117b7a359f5fec49478fd9d7f50f996a7ef53 – there should be explicitly listed what are you renaming. Either something along `KA-242 Rename ItemListController and ItemUrlGeneratorTest` or even better: 
```
KA-242 Rename few classes

ItemListController → ItemsController
ItemListControllerTest → ItemsControllerTest
ItemUrlGeneratorTest → UrlGeneratorTest
````
* b67d172809dab83071579504dfa511252ae4834f – this should be part of 965912dd343285894428190ff51774cd39f76e71. Commits should be atomic in the feature-wise sense but complete in the code-correctness sense (build should success and tests pass after each commit). Note: On contrary, you perform two separable deads in 155bf5753be30eb083e51712ce14d3043cbec538, for example.

I basically want you to stick up with MS naming conventions and update R# code templates whenever you decide to change your code style. I also don't feel tests are in their final form just yet. Otherwise, we are all good though and I'm sure the task will be finished soon 😺 

(Note: read "outdated" comments as always, please)

LindaLangerova committed:
KA-242 Update Resharper code styling

LindaLangerova committed:
KA-242 Add Async suffixes to repository methods' names

LindaLangerova committed:
KA-242 Synchonize test names with their class names

LindaLangerova committed:
KA-242 Rename AnonymousArg to GetAnonymousArg, move private method down and update it

LindaLangerova committed:
KA-242 Compare arg in test with already existing item

LindaLangerova committed:
KA-242 Apply code cleanup

PetrSvirak commented on TodoApp/src/TodoApp.Api/Controllers/ItemsController.cs:
```diff
@@ -22,30 +22,30 @@ public ItemsController(IItemRepository repository, IUrlGenerator urlGenerator)
         public async Task<IHttpActionResult> DeleteAsync(Guid id)
         {
-            _repository.Delete(id);
+             _repository.Delete(id);
```
One never knows when they will require extra space, right 😛

PetrSvirak commented on TodoApp/src/TodoApp.Contract/Repositories/IItemRepository.cs:
```diff
@@ -1,15 +1,16 @@
-        Item Get(Guid id);
-        Item Update(Guid id, Item item);
-        Item Add(Item item);
+        Task<Item[]> GetAll();
```
I liked `IEnumerable<Item>` more, why have you changed the type?

LindaLangerova replied:
Well, I guess I had some problems with type conversions when there was IEnumerable, but I' not really sure now... 😟 Actually, in task 3, I converted it to List 😄 do you think IEnumerable is still better than List? Or that List is ok? 😇

PetrSvirak replied:
As always, it depends. There are several aspects mostly connected with the content of the collection and usage. But in general:
* you want to avoid [premature materialization](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/intermediate-materialization)
* you want to control what your interface users do with the data (returning list gives an easy way to `Add` some entities (after the `GetAll` method is called) and indexed access) and returning general interfaces makes it harder to unintentionally spoil a collection

Let's discuss this in person as well 😃

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/TodoApp.Api.Tests.csproj:
```diff
@@ -86,9 +86,9 @@
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="Services\UrlGenerator.test.cs" />
+    <Compile Include="Services\UrlGeneratorTests.test.cs" />
```
Still having the suffix `.test.cs` instead of `.cs` 😉

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/UrlGeneratorTests.test.cs:
```diff
@@ -35,5 +29,15 @@ public void GetItemUrl_UrlReceived()
             Assert.That(receivedUrl, Is.EqualTo(requestedUrl));
         }
+
+        private static bool ContainsCorrectId(object first)
```
Great you did not name the function `GetAnonymousArg` as commit message states 😉

PetrSvirak replied:
I don't like `first` parameter name - choose a more descriptive one - how is the second (😉) argument called in the `urlHelper.Route` method? That is the argument you resolve upon, isn't it?

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Services/UrlGeneratorTests.test.cs:
```diff
@@ -35,5 +29,15 @@ public void GetItemUrl_UrlReceived()
+        {
+            new RouteValueDictionary(first).TryGetValue("id", out var firstId);
+
+            if (firstId != null && firstId.Equals(Guid.Parse("5f6a2723-040a-4398-8b63-9d55153378ba")))
```
Since you use this string with a GUID (or even better: the whole GUID) multiple times, it would be great to create a constant (or a static read-only field) for it...

PetrSvirak replied:
The `TryGetValue` returns a boolean that determines whether some `value` was found under given `key`. You can simply call `firstId* == Guid.Parse(…)` and omit the `null-check` - this combined could make your method more readable.

* - `firstId` is not a nice name - `id` or `itemId` would be a bit more descriptive...

PetrSvirak commented on TodoApp/test/TodoApp.Api.Tests/Controllers/ItemsControllerTests.test.cs:
```diff
@@ -113,8 +113,7 @@ public void SetUp()
             _repository.UpdateAsync(id,
-                               Arg.Is<Item>(item => item.Id.ToString() == "c5cc89a0-ab8d-4328-9000-3da679ec02d3"
-                                                    && item.Text == "Make a coffee"))
+                               Arg.Is<Item>(item => item.Equals(updateItem) ))
```
Unless I'm mistaken, you are using reference comparison here, meaning your `updateItem` is passed through the controller to the repository. This is a bit fragile, but I would rather discuss it in person than write an essay.

PetrSvirak reviewed (approved):
Your commits got a bit unorganized again 😉 But it is mostly good. I would like to discuss some details in persons - especially go deeper with You on formatting possibilities in sense readability and object comparison in tests, please, do definitely remind me these on closest code-involved occasion.

Go through my comments (including outdated) below, please. There are some more details in your code that should be fixed, however, I believe they can be solved within the following task if You agree (express so by merging the PR 😉; otherwise, feel free to continue with commits).
